client/chat/channel.tsx

```diff
@@ -463,109 +461,66 @@ class Channel extends React.Component<ChannelProps> {
 
-interface ChatChannelViewProps {
-  user: SelfUserRecord
-  chat: ReturnType<typeof ChatState>
-  dispatch: DispatchFunction<ReduxAction>
-  params: { channel: string }
-}
+  const prevChannelName = usePrevious(channelName)
+  const prevChannel = usePrevious(channel)
 
-function isLeavingChannel(oldProps: ChatChannelViewProps, newProps: ChatChannelViewProps) {
-  return (
-    oldProps.params.channel.toLowerCase() === newProps.params.channel.toLowerCase() &&
-    oldProps.chat.byName.has(oldProps.params.channel.toLowerCase()) &&
-    !newProps.chat.byName.has(oldProps.params.channel.toLowerCase())
-  )
-}
-
-class ChatChannelView extends React.Component<ChatChannelViewProps> {
-  componentDidMount() {
-    const routeChannel = decodeURIComponent(this.props.params.channel)
-    if (this._isInChannel()) {
-      this.props.dispatch(retrieveUserList(routeChannel))
-      this.props.dispatch(retrieveInitialMessageHistory(routeChannel))
-      this.props.dispatch(activateChannel(routeChannel) as any)
-    } else {
-      this.props.dispatch(joinChannel(routeChannel))
-    }
-  }
-
-  componentDidUpdate(prevProps: ChatChannelViewProps) {
-    if (isLeavingChannel(prevProps, this.props)) {
+  // TODO(2Pac): Pull this out into some kind of "isLeaving" hook and share with whispers/lobby?
+  useEffect(() => {
+    if (prevChannelName === channelName && prevChannel && !channel) {
       push('/')
-      return
     }
+  })
 
-    const prevChannel = decodeURIComponent(prevProps.params.channel)
-    const routeChannel = decodeURIComponent(this.props.params.channel)
-    if (this._isInChannel()) {
-      this.props.dispatch(retrieveUserList(routeChannel))
-      this.props.dispatch(retrieveInitialMessageHistory(routeChannel))
-      this.props.dispatch(activateChannel(routeChannel) as any)
-    } else if (
-      !prevProps.chat.byName.has(routeChannel) &&
-      prevChannel.toLowerCase() !== routeChannel.toLowerCase()
-    ) {
+  // TODO(2Pac): Make this less spammy?
+  useEffect(() => {
+    dispatch(activateChannel(channelName) as any)
+  }, [channel])
+
+  useEffect(() => {
+    if (channel) {
+      dispatch(retrieveUserList(channelName))
+      dispatch(retrieveInitialMessageHistory(channelName))
+    } else {
       if (MULTI_CHANNEL) {
-        this.props.dispatch(joinChannel(routeChannel))
+        dispatch(joinChannel(channelName))
       } else {
         push('/')
       }
     }
-    if (prevChannel && prevChannel.toLowerCase() !== routeChannel.toLowerCase()) {
-      this.props.dispatch(deactivateChannel(prevChannel) as any)
-    }
-  }
 
-  componentWillUnmount() {
-    this.props.dispatch(deactivateChannel(decodeURIComponent(this.props.params.channel)) as any)
-  }
-
-  render() {
-    const routeChannel = decodeURIComponent(this.props.params.channel)
-    const channel = this.props.chat.byName.get(routeChannel.toLowerCase())
+    return () => dispatch(deactivateChannel(channelName) as any)
```

Comment:
Am I missing something in how `useEffect` works, but I thought it would call the cleanup function each time after it executes the effect? And if activating the channel depends on the changes to `channel`, putting a cleanup function which deactivates the channel each time it's run into the same effect will just cause an infinite loop.

Seems to me like conceptually we're deactivating channel at the right times, e.g. when navigating away from a channel (either when a channel name in route params changes, or the component unmounts), but for channel activation we're not doing the same (well, symmetrical) thing. I think this is one of those cases where react's hook usage actually brings some underlying issues up to the surface. So, yay hooks? :d

I feel like we _should_ be able to write this hook:
```js
useEffect(() => {
  dispatch(activateChannel(channelName))

  return () => dispatch(deactivateChannel(channelName))
}, [channelName])
```
But currently we're activating the channel each time its record in the store changes, and I'm not sure why.